src/main/java/com/redhat/devtools/lsp4ij/settings/jsonSchema/LSPServerConfigurationJsonSchemaFileProvider.java

```diff
@@ -0,0 +1,84 @@
+package com.redhat.devtools.lsp4ij.settings.jsonSchema;
+
+import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.application.WriteAction;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.impl.PsiManagerEx;
+import com.intellij.psi.impl.file.impl.FileManagerImpl;
+import com.intellij.util.ModalityUiUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+public class LSPServerConfigurationJsonSchemaFileProvider extends AbstractLSPJsonSchemaFileProvider {
+
+    private static final Object SCHEMA_CHANGE_REQUESTOR = new Object();
+
+    private final int index;
+    private final @NotNull Project project;
+    private MyLightVirtualFile file;
+    private boolean unused;
+
+    public LSPServerConfigurationJsonSchemaFileProvider(int index, @NotNull Project project) {
+        super(generateJsonFileName(index));
+        this.index = index;
+        this.project = project;
+        this.file = new MyLightVirtualFile("lsp.server." + index + ".schema.json", "");
+        this.unused = true;
+    }
+
+    private static String generateJsonFileName(int index) {
+        return "lsp.server." + index + ".json";
+    }
+
+    public int getIndex() {
+        return index;
+    }
+
+    @Override
+    public @Nullable VirtualFile getSchemaFile() {
+        return file;
+    }
+
+    public void reset() {
+        try {
+            updateFileContent("", file, project);
+        } finally {
+            unused = true;
+        }
+    }
+
+    public void setSchemaContent(@NotNull String jsonSchemaContent) {
+        try {
+            updateFileContent(jsonSchemaContent, file, project);
+        } finally {
+            unused = false;
+        }
+    }
+
+    public boolean isUnused() {
+        return unused;
+    }
+
+
+    private static void updateFileContent(@NotNull String content,
+                                          @NotNull MyLightVirtualFile file ,
+                                          @NotNull Project project) {
+        if (Objects.equals(content, file.getContent())) {
+            return;
+        }
+        file.setContent(SCHEMA_CHANGE_REQUESTOR, content, false);
+        reloadPsi(file, project);
+    }
+
+    private static void reloadPsi(final VirtualFile file, final Project project) {

```

Comment:
I'm a pretty solid stickler for using JetBrains' nullability annotations consistently for params/return values so that, at a glance, I know when/whether it's okay to dereference something. It looks like both `file` and `project` are expected to be non-null, so it might make sense to annotate them as `@NotNull` to be explicit about that.

This same comment would hold just about anywhere that inputs/outputs aren't currently expressing their nullability expectations.

Reply:
I have just translate this code https://github.com/JetBrains/intellij-community/blob/f873c09975f3262558093855d04474598356d2a5/platform/lang-impl/src/com/intellij/openapi/roots/impl/PushedFilePropertiesUpdaterImpl.kt#L515 but I agree we need to do that.